pyriemann/spatialfilters.py

```diff
@@ -1,17 +1,203 @@
+            the percentage specified by ``val``.
+            ``val`` must be a float in (0,1], typically 0,99.
+        If ``{'max_cond': val}`` :
+            dimension reduction selecting the number of components such that
+            the condition number of the mean covariance matrix is lower than
+            ``val``.
+            This threshold has a physiological interpretation, because it can
+            be viewed as the ratio between the power of the strongest component
+            (usually, blink source) and the power of the lowest component you
+            don't want to keep (acquisition sensor noise).
+            ``val`` must be a float strictly superior to 1, typically 100.
+    verbose : bool (default False)
+        Verbose flag.
+
+    Attributes
+    ----------
+    n_components_ : int
+        If fit, the number of components after dimension reduction.
+    filters_ : ndarray, shape ``(n_channels_, n_components_)``
+        If fit, the spatial filters to whiten covariance matrices.
+    inv_filters_ : ndarray, shape ``(n_components_, n_channels_)``
+        If fit, the spatial filters to unwhiten covariance matrices.
+
+    Notes
+    -----
+    .. versionadded:: 0.2.7.dev
+
+    """
+
+    def __init__(self, metric='euclid', dim_red=None, verbose=False):
+        """Init."""
+        self.metric = metric
+
+        if dim_red is None:
+            pass
+        elif isinstance(dim_red, dict):
+            if len(dim_red) > 1:
+                raise ValueError(
+                    'Dictionary dim_red must contain only one element (Got %d)'
+                    % len(dim_red))
+
+            self._dr_key = next(iter(dim_red))
+            self._dr_val = dim_red.get(self._dr_key)
+            if self._dr_key == 'n_components':
+                if self._dr_val < 1:
+                    raise ValueError(
+                        'Value n_components must be superior to 1 (Got %d)'
+                        % self._dr_val)
+                if not isinstance(self._dr_val, numbers.Integral):
+                    raise ValueError(
+                        'n_components=%d must be of type int (Got %r)'
+                        % (self._dr_val, type(self._dr_val)))
+            elif self._dr_key == 'expl_var':
+                if not 0 < self._dr_val <= 1:
+                    raise ValueError(
+                        'Value expl_var must be included in (0, 1] (Got %d)'
+                        % self._dr_val)
+            elif self._dr_key == 'max_cond':
+                if self._dr_val <= 1:
+                    raise ValueError(
+                        'Value max_cond must be strictly superior to 1 '
+                        '(Got %d)' % self._dr_val)
+            else:
+                raise ValueError(
+                    'Unknown key in parameter dim_red: %r' % self._dr_key)
+        else:
+            raise ValueError('Unknown type for parameter dim_red')
```

Comment:
all this parameter validation should be done in fit and not in the init. This violates the scikit-learn API.
The init should not add any other attribute that the params passed to the init.
If someone uses set_params from BaseEstimator after the init you will have a problem otherwise.

Reply:
You are right: I will test parameters in `fit` rather than in `_init_`.